swift/schema.py

```diff
@@ -449,9 +449,36 @@ class KeyPathApplicationExpr(Expr):
 class KeyPathDotExpr(Expr):
     pass
 
+class KeyPathComponent(AstNode):
+    """
+    A component of a `KeyPathExpr`.
+    """
+    kind: int | doc("kind of key path component") | desc("""
+        This is 3 for properties, 4 for array and dictionary subscripts, 5 for optional forcing
+        (`!`), 6 for optional chaining (`?`), 7 for implicit optional wrapping, 8 for `self`,
+        and 9 for tuple element indexing.
+
+        The following values should not appear: 0 for invalid components, 1 for unresolved
+        properties, 2 for unresolved subscripts, 10 for #keyPath dictionary keys, and 11 for
+        implicit IDE code completion data.
+    """)
```

Comment:
Instead of exposing this `kind` integer directly, could we add `is_property : predicate`, `is_subscript : predicate`, etc.? This leaves us with the freedom to change the numbering without fear of breaking any QL code.

Reply:
> I discussed this with @redsun82; we hadn't considered the option of predicates specifically, but did think about making AST subclasses for the different types of component. The rationale for sticking with the int was to minimise the amount of work in the extractor (i.e. not to introduce a switch statement) and just emit the int we get from the Swift AST directly. For comparison, `ObjectLiteralExpr` [follows the same pattern](https://github.com/github/codeql/blob/b7c7449b087a73e06cf5486f8c0aa415feeea441/swift/schema.py#L678).
> 
> What do you think?

I'm a big of Keeping the Extractor Stupid, and I do see that our usage of `: predicate` isn't super consistent.

I'm fine with either approach, really. Adding AST subclasses for the different types of components probably achieves the same goal: making sure users don't depend on the exact numbering. If we do keep the kind exposed in QL we should preface the generated QLDoc with a big fat "INTERNAL: Do not use" comment.

> If we do migrate to predicates, would you propose separate predicates for each of the unexpected kinds (`Invalid`, `Unresolved*`, `DictionaryKey`, `CodeCompletion`), or just lump them all in as "invalid"? And what about non-enumerated kinds like `static_cast<KeyPath::Component::Kind>(123456789)` if we run into them?

Good question! I think it would make sense lump them all into an "invalid"/"unsupported" case with a default in the hypothetical switch statement.

Reply:
Let's keep the int with an "INTERNAL: Do not use" comment. Would that go in `desc` or `doc`? Feel a bit long for `doc`, but maybe `desc` isn't prominent enough? @redsun82 was talking about adding codegen support for internal fields soon-ish, so maybe it's not a huge deal either way.

And then the predicates would go in the "stub" in [swift/ql/lib/codeql/swift/elements/KeyPathComponent.qll](https://github.com/github/codeql/blob/b900185ae30d3ab65d80359f080f4eb65058f404/swift/ql/lib/codeql/swift/elements/KeyPathComponent.qll)?

Reply:
Oh, I thought I answered this one. Sorry about that. I think the QLDoc added in 64443dfdcf320a7e24fe6cc7c984bc97bbafc9d5 is perfect 👍.

And yes, I think adding the predicates to https://github.com/github/codeql/blob/b900185ae30d3ab65d80359f080f4eb65058f404/swift/ql/lib/codeql/swift/elements/KeyPathComponent.qll is the most appropriate place.